### DISCUSSION TITLE
🐼 Panda v0.38.0

### DISCUSSION THREAD

**Discussion by anubra266** [chakra-ui/panda#2543] [Announcements]
To upgrade to v0.38.0:

```bash
pnpm up "@pandacss/*"
#or
yarn up '@pandacss/*'
```

### Fixed

- Fix css reset regressions where:
  - first letter gets a different color
  - input or select gets a default border
- Fix type import
- Fix Panda imports detection when using `tsconfig`.`baseUrl` with an outdir that starts with `./`.

### Added

- Add least resource used (LRU) cache in the hot parts to prevent memory from growing infinitely
- Add support for deprecating tokens, utilities, patterns and config recipes.

Set the `deprecated` property to `true` to enable deprecation warnings. Alternatively, you can also set it to a string
to provide a custom migration message.

**Deprecating a utility**

```js
defineConfig({
  utilities: {
    ta: {
      deprecated: true,
      transform(value) {
        return { textAlign: value }
      },
    },
  },
})
```

**Deprecating a token**

```js
defineConfig({
  theme: {
    tokens: {
      spacing: {
        lg: { value: '8px', deprecated: 'use `8` instead' },
      },
    },
  },
})
```

**Deprecating a pattern**

```js
defineConfig({
  patterns: {
    customStack: {
      deprecated: true,
    },
  },
})
```

**Deprecating a recipe**

```js
defineConfig({
  theme: {
    recipes: {
      btn: {
        deprecated: 'will be removed in v2.0',
      },
    },
  },
})
```

- Add support for array values in the special `css` property for the JSX factory and JSX patterns

This makes it even easier to merge styles from multiple sources.

```tsx
import { Stack, styled } from '../styled-system/jsx'

const HeroSection = (props) => {
  return (
    <Stack css={[{ color: 'blue.300', padding: '4' }, props.css]}>
      <styled.div css={[{ fontSize: '2xl' }, props.hero]}>Hero Section</styled.div>
    </Stack>
  )
}

const App = () => {
  return (
    <>
      <HeroSection css={{ backgroundColor: 'yellow.300' }} hero={css.raw({ fontSize: '4xl', color: 'red.300' })} />
    </>
  )
}
```

should render something like:

```html
<div class="d_flex flex_column gap_10px text_blue.300 p_4 bg_yellow.300">
  <div class="fs_4xl text_red.300">Hero Section</div>
</div>
```

**Comment by antonjoel82** [comment]
@anubra266 Awesome, thanks for all these changes!

To make sure I understand -- `css` accepting arrays allows it to better handle merging styles (objects or CSS template literals) while `cx` is still just used for merging the resulting `className` / `string` output of a `css(...)` invocation (or other classnames)?

**Reply by astahmer** [reply]
exactly ! `css={[xxx, yyy, zzz]}` is equivalent to `css(xxx, yyy, zzz)` which means `xxx` / `yyy` / `zzz` are style objects
whereas `cx(aaa, bbb, ccc)` should be classNames (produced either by `css` or any CSS className really)

**Comment by kiruushaaa** [comment]
@anubra266 @astahmer hi guys! There is an issue appeared when using `cssProp` from `splitCssProps()` https://panda-css.com/docs/concepts/style-props#making-your-own-styled-components

The reproducing actually is an example from the docs 🤓 https://play.panda-css.com/3eM_vCiQv0

ps. `0.38.0` only

**Reply by astahmer** [reply]
will be fixed with https://github.com/chakra-ui/panda/pull/2549

**Reply by astahmer** [reply]
you can explicitly use `...(Array.isArray(cssProp) ? cssProp : [cssProp])` in the meantime
